lib/windows/launcher/actions/desktopShortcutActions.js

```diff
@@ -427,71 +429,40 @@ function createShortcutForMacOS(app) {
-                                        fs.chmodSync(appExecPath, '755');
-                                        return null;
-                                    });
-                                filePath = path.join(config.getHomeDir(), `/Applications/${fileName}.app`);
-                                ncp(
-                                    tmpAppTemplatePath,
-                                    filePath,
-                                    errCopy => {
-                                        if (errCopy) {
-                                            return dispatch(ErrorDialogActions.showDialog(`Fail to create desktop shortcut on MacOS when copying app from tmp folder to application folder with error: ${errCopy}`));
-                                        }
-                                        // Change file to be executable
-                                        const appExecPath = path.join(filePath, '/Contents/MacOS/Application Stub');
-                                        fs.chmodSync(appExecPath, '755');
-                                        return null;
-                                    });
-                                return null;
-                            },
-                        );
-                        return null;
-                    });
-                return null;
-            });
-        return null;
+        fileUtil.chmodDir(tmpAppTemplatePath, mode)
+        .then(() => fileUtil.copy(appTemplatePath, tmpAppTemplatePath))
+        .then(() => fileUtil.chmodDir(tmpAppTemplatePath, mode))
+        .then(() => new Promise((resolve, reject) => {
+            // Convert icns file for the app
+            const icnsPath = path.join(tmpAppTemplatePath, '/Contents/Resources/icon.icns');
+            const input = fs.readFileSync(app.iconPath);
+            const output = png2icons.PNG2ICNS(input, png2icons.HERMITE, false, 0);
+            if (output) {
+                fs.writeFileSync(icnsPath, output);
+                resolve();
+            }
+            reject();
+        }))
+        .then(() => fileUtil.createTextFile(wflowPath, wflowContent))
+        .then(() => fileUtil.createTextFile(infoPath, infoContent))
+        .then(() => fileUtil.copy(tmpAppTemplatePath, filePath))
+        .then(() => fileUtil.chmodDir(filePath, mode))
+        .then(() => new Promise(resolve => {
+            const appExecPath = path.join(filePath, '/Contents/MacOS/Application Stub');
+            fs.chmodSync(appExecPath, mode);
+            filePath = path.join(config.getHomeDir(), `/Applications/${fileName}.app/`);
+            resolve();
+        }))
+        .then(() => fileUtil.copy(tmpAppTemplatePath, filePath))
+        .then(() => fileUtil.chmodDir(filePath, mode))
+        .then(() => new Promise(resolve => {
+            const appExecPath = path.join(filePath, '/Contents/MacOS/Application Stub');
+            fs.chmodSync(appExecPath, mode);
+            resolve();
```

Comment:
Why is this being done twice? Already done on line 450-454.

Also, this could be written as:

```
.then(() => {
    const appExecPath = path.join(filePath, '/Contents/MacOS/Application Stub');
    return fs.chmod(appExecPath, mode);
}
```

This is given that we use the `fs-extra` module instead of `fs`. It is a drop in replacement for `fs` that promisifies the functions from `fs`.

Reply:
Yes, I will modify here.

Reply:
This is being done twice since on Mac there are desktop and launchpad, I create launcher for both desktop and launchpad.